diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 2ce03ed..d2f237a 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -2288,11 +2288,26 @@ out_err:
 
 static int nl80211_mgmt_subscribe_ap_dev_sme(struct i802_bss *bss)
 {
+        static const int stypes[] = {
+               WLAN_FC_STYPE_AUTH,
+               WLAN_FC_STYPE_ASSOC_REQ,
+               WLAN_FC_STYPE_REASSOC_REQ,
+        };
+        unsigned int i;
+
 	if (nl80211_alloc_mgmt_handle(bss))
 		return -1;
 	wpa_printf(MSG_DEBUG, "nl80211: Subscribe to mgmt frames with AP "
 		   "handle %p (device SME)", bss->nl_mgmt);
 
+        for (i = 0; i < ARRAY_SIZE(stypes); i++) {
+                if (nl80211_register_frame(bss, bss->nl_mgmt,
+                                           (WLAN_FC_TYPE_MGMT << 2) |
+                                           (stypes[i] << 4),
+                                           NULL, 0) < 0) {
+                         goto out_err;
+                }
+        }
 	if (nl80211_action_subscribe_ap(bss))
 		goto out_err;
 
