Index: ntpclient-2010/ntpclient.c
===================================================================
--- ntpclient-2010.orig/ntpclient.c
+++ ntpclient-2010/ntpclient.c
@@ -127,6 +127,7 @@ struct ntp_control {
 	int probe_count;
 	int cycle_time;
 	int goodness;
+	int timezone;
 	int cross_check;
 	char serv_addr[4];
 };
@@ -169,14 +170,14 @@ static int set_freq(int new_freq)
 #endif
 }
 
-static void set_time(struct ntptime *new)
+static void set_time(struct ntptime *new, int timezone)
 {
 #ifndef USE_OBSOLETE_GETTIMEOFDAY
 	/* POSIX 1003.1-2001 way to set the system clock
 	 */
 	struct timespec tv_set;
 	/* it would be even better to subtract half the slop */
-	tv_set.tv_sec  = new->coarse - JAN_1970;
+	tv_set.tv_sec  = new->coarse - JAN_1970 + (60 * timezone);
 	/* divide xmttime.fine by 4294.967296 */
 	tv_set.tv_nsec = USEC(new->fine)*1000;
 	if (clock_settime(CLOCK_REALTIME, &tv_set)<0) {
@@ -191,7 +192,7 @@ static void set_time(struct ntptime *new
 	 */
 	struct timeval tv_set;
 	/* it would be even better to subtract half the slop */
-	tv_set.tv_sec  = new->coarse - JAN_1970;
+	tv_set.tv_sec  = new->coarse - JAN_1970 + (60 * timezone);
 	/* divide xmttime.fine by 4294.967296 */
 	tv_set.tv_usec = USEC(new->fine);
 	if (settimeofday(&tv_set,NULL)<0) {
@@ -401,7 +402,7 @@ static int rfc1305print(u32 *data, struc
 
 	/* XXX should I do this if debug flag is set? */
 	if (ntpc->set_clock) { /* you'd better be root, or ntpclient will exit here! */
-		set_time(&xmttime);
+		set_time(&xmttime, ntpc->timezone);
 	}
 
 	/* Not the ideal order for printing, but we want to be sure
@@ -587,6 +588,7 @@ static void usage(char *argv0)
 #endif
 	" [-f frequency] [-g goodness] -h hostname\n"
 	"\t[-i interval] [-l] [-p port] [-q min_delay]"
+	" [-z timezone]"
 #ifdef ENABLE_REPLAY
 	" [-r]"
 #endif
@@ -611,7 +613,7 @@ int main(int argc, char *argv[]) {
 	ntpc.cross_check=1;
 
 	for (;;) {
-		c = getopt( argc, argv, "c:" DEBUG_OPTION "f:g:h:i:lp:q:" REPLAY_OPTION "stD");
+		c = getopt( argc, argv, "c:" DEBUG_OPTION "f:g:h:i:lp:q:z:" REPLAY_OPTION "stD");
 		if (c == EOF) break;
 		switch (c) {
 			case 'c':
@@ -646,6 +648,9 @@ int main(int argc, char *argv[]) {
 			case 'q':
 				min_delay = atof(optarg);
 				break;
+			case 'z':
+				ntpc.timezone = atoi(optarg);
+				break;
 #ifdef ENABLE_REPLAY
 			case 'r':
 				do_replay();
@@ -693,11 +698,12 @@ int main(int argc, char *argv[]) {
 		"  -l live        %d\n"
 		"  -p local_port  %d\n"
 		"  -q min_delay   %f\n"
+		"  -z timezone	  %d\n"
 		"  -s set_clock   %d\n"
 		"  -x cross_check %d\n",
 		ntpc.probe_count, debug, ntpc.goodness,
 		hostname, ntpc.cycle_time, ntpc.live, udp_local_port, min_delay,
-		ntpc.set_clock, ntpc.cross_check );
+		ntpc.timezone, ntpc.set_clock, ntpc.cross_check );
 	}
 
 	/* Startup sequence */
