--- a/drivers/net/wireless/ath/ath10k/debug.h
+++ b/drivers/net/wireless/ath/ath10k/debug.h
@@ -43,6 +43,7 @@ enum ath10k_debug_mask {
 	ATH10K_DBG_USB		= 0x00040000,
 	ATH10K_DBG_USB_BULK	= 0x00080000,
 	ATH10K_DBG_SMART_ANT    = 0x01000000,
+	ATH10K_DBG_FWLOG        = 0x10000000,
 	ATH10K_DBG_ANY		= 0xffffffff,
 };
 
--- a/drivers/net/wireless/ath/ath10k/fwlog.c
+++ b/drivers/net/wireless/ath/ath10k/fwlog.c
@@ -1471,8 +1471,8 @@ static void ath10k_fwlog_parse_msg(struc
 	len -= sizeof(dropped);
 
 	if (dropped > 0)
-		printk(FW_DBGLOG_PRINT_PREFIX "%d log buffers are dropped \n",
-		       dropped);
+		ath10k_dbg(ar, ATH10K_DBG_FWLOG, FW_DBGLOG_PRINT_PREFIX
+			   "%d log buffers are dropped \n\n", dropped);
 
 	count = 0;
 	buffer = (u32 *)data;
