--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -1389,15 +1389,12 @@ static ssize_t ath10k_write_fw_dbglog(st
 {
 	struct ath10k *ar = file->private_data;
 	int ret;
-	char buf[96];
+	char buf[96] = {0};
 	unsigned int log_level;
 	u64 mask;
 
 	simple_write_to_buffer(buf, sizeof(buf) - 1, ppos, user_buf, count);
 
-	/* make sure that buf is null terminated */
-	buf[sizeof(buf) - 1] = 0;
-
 	ret = sscanf(buf, "%llx %u", &mask, &log_level);
 
 	if (!ret)
