Index: fstools-2016-01-10/jffs2reset.c
===================================================================
--- fstools-2016-01-10.orig/jffs2reset.c
+++ fstools-2016-01-10/jffs2reset.c
@@ -119,7 +119,7 @@ int main(int argc, char **argv)
 		return -1;
 	}
 
-	v = volume_find("rootfs_data");
+	v = volume_find("overlay_volume");
 	if (!v) {
 		ULOG_ERR("MTD partition 'rootfs_data' not found\n");
 		return -1;
